### PR TITLE
feat(context): improve @inject.setter and add @inject.binding

### PR DESCRIPTION
**Pull request by raymondfeng**
Extracted from https://github.com/strongloop/loopback-next/pull/2635

1. Allow `@inject.setter` to set other forms of value providers than constant

```ts
// Set the binding to a constant value, this is the current behavior
injectedSetter('my-value'); 

// or use the returned binding to configure it
const binding = injectedSetter();
binding.toClass(MyClass);
```

2. Allows `bindingCreation` option to control how the underlying binding is resolved/created.

## Checklist

👉 [Read and sign the CLA (Contributor License Agreement)](https://cla.strongloop.com/agreements/strongloop/loopback-next) 👈

- [x] `npm test` passes on your machine
- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [x] API Documentation in code was updated
- [x] Documentation in [/docs/site](../tree/master/docs/site) was updated
- [ ] Affected artifact templates in `packages/cli` were updated
- [ ] Affected example projects in `examples/*` were updated

### CODE REVIEW

File: docs/site/Decorators_inject.md [hunk 1]
```diff
@@ -143,6 +143,66 @@ export class HelloController {
...
+```
+
+```ts
+const binding = this.greetingSetter('Greetings!');
+```
+
+or
+
+```ts
+const binding = this.greetingSetter().toDynamicValue(() => 'Greetings!');
```
Comment (bajtos) [inline docs/site/Decorators_inject.md]: @raymondfeng could you please remind me and other people reviewing this pull request why are we adding this feature? What use cases are you trying to support?

File: packages/authentication/src/__tests__/unit/providers/authentication.provider.unit.ts [hunk 2]
```diff
@@ -110,7 +110,10 @@ describe('AuthenticateActionProvider', () => {
       strategy.setMockUser(mockUser);
       provider = new AuthenticateActionProvider(
         () => Promise.resolve(strategy),
-        u => (currentUser = u),
+        (...u) => {
```
Comment (bajtos) [inline packages/authentication/src/__tests__/unit/providers/authentication.provider.unit.ts]: I don't see how is this change related to the enhanced setter. Could you please explain the reasoning, or revert this part back?

File: packages/context/src/__tests__/acceptance/class-level-bindings.acceptance.ts [hunk 3]
```diff
@@ -17,6 +17,7 @@ import {
   ResolutionSession,
   Setter,
 } from '../..';
+import {BindingCreationPolicy} from '../../context';
```
Comment (bajtos) [inline packages/context/src/__tests__/acceptance/class-level-bindings.acceptance.ts]: Please import directly from `../..` and merge the import with the lines above.

File: packages/context/src/__tests__/unit/context.unit.ts [hunk 4]
```diff
@@ -12,6 +12,7 @@ import {
   Context,
   isPromiseLike,
 } from '../..';
+import {BindingCreationPolicy} from '../../context';
```
Comment (bajtos) [inline packages/context/src/__tests__/unit/context.unit.ts]: Ditto, please import from `../..`

File: packages/context/src/context.ts [hunk 5]
```diff
@@ -737,6 +741,30 @@ export class Context extends EventEmitter {
...
+    key: BindingAddress<T>,
+    policy?: BindingCreationPolicy,
+  ) {
+    let binding: Binding<T>;
+    if (policy === BindingCreationPolicy.ALWAYS_CREATE) {
+      binding = this.bind(key);
+    } else if (policy === BindingCreationPolicy.NEVER_CREATE) {
+      binding = this.getBinding(key);
+    } else {
+      if (this.isBound(key)) {
```
Comment (bajtos) [inline packages/context/src/context.ts]: Let's reduce nesting:

```ts
} else if (this.isBound(key)) {
```

File: packages/context/src/inject.ts [hunk 6]
```diff
@@ -188,7 +188,9 @@ export function inject(
 /**
  * The function injected by `@inject.getter(bindingSelector)`.
  */
-export type Getter<T> = () => Promise<T>;
+export interface Getter<T> {
+  (): Promise<T>;
+}
```
Comment (bajtos) [inline packages/context/src/inject.ts]: What is the reasoning for this change? Do you need an interface so that you can inherit from it?

File: packages/context/src/inject.ts [hunk 7]
```diff
@@ -201,9 +203,32 @@ export namespace Getter {
 }
 
 /**
- * The function injected by `@inject.setter(key)`.
+ * The function injected by `@inject.setter(bindingKey)`.
  */
-export type Setter<T> = (value: T) => void;
+export interface Setter<T> {
```
Comment (bajtos) [inline packages/context/src/inject.ts]: What is the reasoning for this change? Do you need an interface so that you can inherit from it?

Also the API docs seems little weird to me - it says `the function` but describes an interface.

File: packages/context/src/inject.ts [hunk 8]
```diff
@@ -351,15 +376,22 @@ function resolveAsSetter(ctx: Context, injection: Injection) {
       `The type of ${targetName} (${targetType.name}) is not a Setter function`,
     );
   }
-  const bindingSelector = injection.bindingSelector;
-  if (!isBindingAddress(bindingSelector)) {
+  const key = injection.bindingSelector;
+  if (!isBindingAddress(key)) {
```
Comment (bajtos) [inline packages/context/src/inject.ts]: Can we keep `bindingSelector` please? I find that name more descriptive than just `key`.

File: packages/authentication/src/__tests__/unit/providers/authentication.provider.unit.ts (hunk 2 above)
Comment (raymondfeng) [reply packages/authentication/src/__tests__/unit/providers/authentication.provider.unit.ts]: It was a left-over and should be reverted now.

File: packages/context/src/inject.ts (hunk 7 above)
Comment (raymondfeng) [reply packages/context/src/inject.ts]: I changed them to interfaces to define `Setter.applyTemplates()`. Now they can be reverted.

File: docs/site/Decorators_inject.md (hunk 1 above)
Comment (raymondfeng) [reply docs/site/Decorators_inject.md]: Added more description.

**Comment by raymondfeng** [comment]
> Personally, I prefer UNIX philosophy - many small tools where each tool does one job and does it well. In that light, I would prefer a new decorator for injecting the entire Binding instance. You have mentioned @inject.binding yourself in one of the earlier pull requests, what was the reasoning for your decision to put this new feature into the existing @inject.setter function?

I'm open here but I would like to avoid a new `@inject.*` flavor before it's necessary. Here is my reasoning:

`@inject.setter` was introduced to bind a `value (getter)` to the underlying key. The 1st version only allows constant value (binding.to()) by calling `setter(constantValue)`. This PR extends the capability to have more control by exposing the Binding instance. The initial version was to allow `BindingTemplate` functions to be accepted by the setter function but we cannot differentiate `constant function value` from `template function`.

If we decide to introduce `@inject.binding`, maybe we should deprecate and finally remove `@inject.setter`. It's similar as preferring `@inject(filterByTag('my-tag')` over `@inject.tag('my-tag')`.

**Comment by raymondfeng** [comment]
@bajtos I went ahead to add `@inject.binding` and leave `@inject.setter` to only set constant values.

**Comment by bajtos** [comment]
> I went ahead to add `@inject.binding` and leave `@inject.setter` to only set constant values.

Awesome 👍 

Here is my point of view:

 - I find it confusing to have a decorator called `@inject.setter` to inject a `Binding` instance.
 - Ideally, IoC related code (like `Binding` instances) should not leak into the actual implementation (the code receiving the dependencies via injection).

With `@inject.setter`, the decorated argument/property receives a generic setter function that can be passed to any 3rd-party code (e.g. a package from npm), the consumer does not need to understand anything about our `Context` and `Binding`.

In that sense, `@inject.binding` can be seen as a kind of an anti-pattern. Having said that, I am ok to add that functionality as an advanced tool for situations where `@inject.setter` is not good enough. Once we learn more about use cases where `@inject.binding` is needed, we can look into ways how to provide more clean API(s) to support those cases.

File: packages/context/src/__tests__/unit/context.unit.ts [hunk 9]
```diff
@@ -342,6 +343,86 @@ describe('Context', () => {
     });
   });
 
+  describe('findOrCreateBinding', () => {
+    describe('bindingCreationPolicy - ALWAYS_CREATE', () => {
```
Comment (bajtos) [inline packages/context/src/__tests__/unit/context.unit.ts]: I think we usually use `context` instead of `describe` to group tests of the same subject but under different conditions.

```suggestion
    context('with ALWAYS_CREATE', () => {
```

This way the tests read like:

- findOrCreateBinding with ALWAYS_CREATE creates a new binding even the key is bound
- findOrCreateBinding with ALWAYS_CREATE creates a new binding if not bound
- ...

File: packages/context/src/__tests__/unit/context.unit.ts [hunk 10]
```diff
@@ -342,6 +343,86 @@ describe('Context', () => {
...
+      it('creates a new binding if not bound', () => {
+        const binding = ctx.findOrCreateBinding(
+          'a-new-key',
+          BindingCreationPolicy.ALWAYS_CREATE,
+        );
+        expect(binding.key).to.eql('a-new-key');
+      });
+    });
+
+    describe('bindingCreationPolicy - NEVER_CREATE', () => {
```
Comment (bajtos) [inline packages/context/src/__tests__/unit/context.unit.ts]: Ditto, use `context` instead of `describe`.

File: packages/context/src/__tests__/unit/context.unit.ts [hunk 11]
```diff
@@ -342,6 +343,86 @@ describe('Context', () => {
...
+        expect(() =>
+          ctx.findOrCreateBinding(
+            'a-new-key',
+            BindingCreationPolicy.NEVER_CREATE,
+          ),
+        ).to.throw(/The key 'a-new-key' is not bound to any value in context/);
+      });
+    });
+
+    describe('bindingCreationPolicy - CREATE_IF_NOT_BOUND', () => {
```
Comment (bajtos) [inline packages/context/src/__tests__/unit/context.unit.ts]: And here too - use `context` instead of `describe`.

File: packages/context/src/__tests__/unit/context.unit.ts [hunk 12]
```diff
@@ -342,6 +343,86 @@ describe('Context', () => {
...
+      it('creates a new binding if the key is not bound', () => {
+        const binding = ctx.findOrCreateBinding(
+          'a-new-key',
+          BindingCreationPolicy.CREATE_IF_NOT_BOUND,
+        );
+        expect(binding.key).to.eql('a-new-key');
+      });
+    });
+
+    describe('default bindingCreationPolicy - CREATE_IF_NOT_BOUND', () => {
```
Comment (bajtos) [inline packages/context/src/__tests__/unit/context.unit.ts]: Ditto, use `context` instead of `describe`.

File: packages/context/src/inject.ts [hunk 13]
```diff
@@ -249,12 +274,20 @@ export namespace inject {
...
   export const setter = function injectSetter(
     bindingKey: BindingAddress,
-    metadata?: InjectionMetadata,
+    metadata?: InjectBindingMetadata,
   ) {
     metadata = Object.assign({decorator: '@inject.setter'}, metadata);
     return inject(bindingKey, metadata, resolveAsSetter);
   };
 
+  export const binding = function injectBinding(
```
Comment (bajtos) [inline packages/context/src/inject.ts]: Please add a tsdoc comment for this new API.

File: packages/authentication/src/__tests__/unit/providers/authentication.provider.unit.ts [hunk 14]
```diff
@@ -110,7 +110,10 @@ describe('AuthenticateActionProvider', () => {
       strategy.setMockUser(mockUser);
       provider = new AuthenticateActionProvider(
         () => Promise.resolve(strategy),
-        u => (currentUser = u),
+        u => {
+          currentUser = u;
+          return new Binding('authentication.currentUser').to(currentUser);
```
[the remark this answers is not in the document]
Comment (raymondfeng) [reply packages/authentication/src/__tests__/unit/providers/authentication.provider.unit.ts]: It's a mock-up return a binding object.

**Comment by bajtos** [comment]
@raymondfeng I don't have bandwidth to review this patch today, but I do want to check your updates before this patch is merged. Please give me few more days.

File: packages/context/src/inject.ts [hunk 15]
```diff
@@ -201,9 +203,31 @@ export namespace Getter {
 }
 
 /**
- * The function injected by `@inject.setter(key)`.
+ * The function injected by `@inject.setter(bindingKey)`. It sets the underlying
+ * binding to a const value. Returns the `Binding` object so that the binding
+ * can be further configured, such as setting it to a class with `toClass()`.
```
Comment (bajtos) [inline packages/context/src/inject.ts]: I am not convinced it's a good idea to return the binding from the setter. I don't want us to leak details about our IoC container (Context/Binding classes) to code receiving dependencies. As I wrote above, I see `Setter` as a concept that's agnostic of DI/IoC and I'd like to preserve that.

For use cases where the developer needs to set more than just the bound value, there is the new `@inject.binding` decorator to give full control over the binding.

I am also concerned about robustness of the proposed solution. It allows developers to invoke the injected setter function without any value provided:

```ts
setter(); 
```

What's the expected outcome? For example, if `T` is `string | undefined`, the call above should be valid and should set the target binding value to `undefined`.

For the reasons outline above, I am against changing the signature of `Setter` function and propose to keep it in the current form.

Comment (raymondfeng) [reply packages/context/src/inject.ts]: I can revert the return type back to `void`. What about the `bindingCreation` option?

Comment (raymondfeng) [reply packages/context/src/inject.ts]: reverted.

**Comment by raymondfeng** [comment]
@bajtos PTAL

**Comment by raymondfeng** [comment]
@bajtos Ping.

**Comment by bajtos** [comment]
> What about the `bindingCreation` option?

This option is configured at DI/IoC level in `@inject.setter` arguments, it makes sense to keep it available. As I see it, it allows the developer that's configuring DI wiring to tweak the behavior of the setter that will be injected.

File: packages/context/src/inject.ts [hunk 16]
```diff
@@ -249,16 +269,49 @@ export namespace inject {
...
+  ) {
+    metadata = Object.assign({decorator: '@inject.binding'}, metadata);
+    return inject(bindingKey, metadata, resolveAsBinding);
+  };
+
   /**
    * Inject an array of values by a tag pattern string or regexp
    *
-   * @example
+   * For example,
```
Comment (bajtos) [inline packages/context/src/inject.ts]:
```suggestion
   * For example:
```
